src/bridge/cockpitcgroupsamples.c

```diff
@@ -21,119 +21,223 @@
+  /* If at max for arch, then unlimited => zero */
+  if (val > 0 && val < G_MAXINT64)
+      cockpit_samples_sample (samples, "cgroup.memory.sw-limit", cgroup, val);
+}
+
+static void
+collect_cpu_v1 (CockpitSamples *samples,
+                int dirfd,
+                const char *cgroup)
+{
+  gint64 val;
+
+  val = read_int64 (dirfd, cgroup, "cpuacct.usage");
+  if (val >= 0 && val < G_MAXINT64)
+    cockpit_samples_sample (samples, "cgroup.cpu.usage", cgroup, val/1000000);
+
+  val = read_int64 (dirfd, cgroup, "cpu.shares");
+  if (val > 0 && val < G_MAXINT64)
+    cockpit_samples_sample (samples, "cgroup.cpu.shares", cgroup, val);
+}
+
+static void
+collect_v2 (CockpitSamples *samples,
+            int dirfd,
+            const char *cgroup)
+{
+  gint64 val;
+
+  /* memory.current: single unsigned value in bytes */
+  val = read_int64 (dirfd, cgroup, "memory.current");
+  if (val >= 0 && val < G_MAXINT64)
+    cockpit_samples_sample (samples, "cgroup.memory.usage", cgroup, val);
+
+  /* memory.max: literally says "max" if there is no limit set, which ends up as "0" after integer conversion;
+   * only create samples for actually limited cgroups */
+  val = read_int64 (dirfd, cgroup, "memory.max");
+  if (val > 0 && val < G_MAXINT64)
+    cockpit_samples_sample (samples, "cgroup.memory.limit", cgroup, val);
+
+  /* same as above for swap */
+  val = read_int64 (dirfd, cgroup, "memory.swap.current");
+  if (val >= 0 && val < G_MAXINT64)
+    cockpit_samples_sample (samples, "cgroup.memory.sw-usage", cgroup, val);
+
+  val = read_int64 (dirfd, cgroup, "memory.swap.max");
+  if (val > 0 && val < G_MAXINT64)
+    cockpit_samples_sample (samples, "cgroup.memory.sw-limit", cgroup, val);
+
+  /* cpu.weight: only exists if cpu controller is enabled; integer in range [1, 10000] */
+  val = read_int64 (dirfd, cgroup, "cpu.weight");
+  if (val > 0 && val < G_MAXINT64)
+    cockpit_samples_sample (samples, "cgroup.cpu.shares", cgroup, val);
+
+  /* cpu.stat: keyed file:
+     usage_usec 50000
+     user_usec 40000
+     system_usec 10000
+     */
+  val = read_prefixed_int64 (dirfd, cgroup, "cpu.stat", "usage_usec ");
+  if (val >= 0 && val < G_MAXINT64)
+    cockpit_samples_sample (samples, "cgroup.cpu.usage", cgroup, val/1000);
 }
 
 static void
 notice_cgroups_in_hierarchy (CockpitSamples *samples,
-                             const gchar *prefix,
-                             void (* collect) (CockpitSamples *, const gchar *, const gchar *))
+                             const char *root_dir,
+                             void (* collect) (CockpitSamples *, int, const char *))
 {
-  const gchar *paths[] = { prefix, NULL };
-  gsize prefix_len;
+  const char *paths[] = { root_dir, NULL };
+  gsize root_dir_len = strlen (root_dir);
   FTSENT *ent;
   FTS *fs;
 
-  prefix_len = strlen (prefix);
-
-  fs = fts_open ((gchar **)paths, FTS_NOCHDIR | FTS_COMFOLLOW, NULL);
+  fs = fts_open ((char **)paths, FTS_NOCHDIR | FTS_COMFOLLOW, NULL);
```

Comment:
1. Cool.  I didn't know this class of functions existed!

2. Ugh.  A bit of an impedance mismatch here, now that we're in the `openat()` world.  That could be fixed with a followup, though.

Reply:
Would be nice if there was a walk function that spits out dirfds indeed, but I'm not aware of one.